File: elixir/two-sum.ex
defmodule Solution do
  @spec two_sum(nums :: [integer], target :: integer) :: [integer]
  def two_sum(nums, target) do
    two_sum(Enum.with_index(nums), target, %{})
  end

  defp two_sum([{v, i} | tail], target, seen) do
    if ii = seen[target - v] do
      [i, ii]
    else
      two_sum(tail, target, Map.put(seen, v, i))
    end
  end
end

File: go/two_sum.go
package main

func twoSum(nums []int, target int) []int {
	a := map[int]int{}
	for i, n := range nums {
		if j, ok := a[target-n]; ok {
			return []int{i, j}
		}
		a[n] = i
	}
	return nil
}
